TrailBlaze/Networking/Interceptor.m

```diff
@@ -0,0 +1,125 @@
+    for (int i = (int) gap; i < numberOfPoints; i+=gap) {
+        [prunedPoints addObject:[routePoints objectAtIndex:i]];
+    }
+    return prunedPoints;
+}
+
++ (void) getBestETAPoint: (NSArray *) allPoints interceptorLocation: (CLLocation *) interceptorLocation runnerLocation: (CLLocation *) runnerLocation completion:(void (^)(MKMapItem *bestPoint, NSError *))completion {
+    
+    NSArray *prunedPoints = [self prunePoints:[self getRemainingRoutePoints:allPoints runnerLocation:runnerLocation] numberOfPoints:10];
+    NSLog(@"🌈🌈🌈🌈🌈%@", prunedPoints);
+    etasDifferences = [NSMutableArray array];
+    etaPointPairs = [[NSMutableDictionary alloc] init];
+    dispatch_queue_t serialQueue = dispatch_queue_create("serialQueue", nil);
+    dispatch_group_t group = dispatch_group_create();
+    
+    MKPlacemark *startInterceptorPlacemark = [[MKPlacemark alloc] initWithCoordinate:CLLocationCoordinate2DMake(interceptorLocation.coordinate.latitude, interceptorLocation.coordinate.longitude)];
+    
+    MKPlacemark *startRunnerPlacemark = [[MKPlacemark alloc] initWithCoordinate:CLLocationCoordinate2DMake(runnerLocation.coordinate.latitude, runnerLocation.coordinate.longitude)];
+    
+    MKMapItem *startInterceptorItem = [[MKMapItem alloc] initWithPlacemark:startInterceptorPlacemark];
+    
+    MKMapItem *startRunnerItem = [[MKMapItem alloc] initWithPlacemark:startRunnerPlacemark];
+
+    for (int i=0; i < prunedPoints.count; i++) {
+        NSLog(@"starting operation %@..", @(i));
+        
+        dispatch_group_enter(group);
+        MKPlacemark *destinationPlacemark = [[MKPlacemark alloc] initWithCoordinate:CLLocationCoordinate2DMake([[prunedPoints objectAtIndex:i][0] doubleValue], [[prunedPoints objectAtIndex:i][1] doubleValue])];
+
+        MKMapItem *destinationItem =  [[MKMapItem alloc] initWithPlacemark:destinationPlacemark];
+
+        MKDirectionsRequest *pathInterceptorRequest = [[MKDirectionsRequest alloc] init];
+        [pathInterceptorRequest setSource:startInterceptorItem];
+        [pathInterceptorRequest setDestination:destinationItem];
+        [pathInterceptorRequest setTransportType:MKDirectionsTransportTypeWalking];
+        [pathInterceptorRequest setRequestsAlternateRoutes:NO];
+        
+        MKDirectionsRequest *pathRunnerRequest = [[MKDirectionsRequest alloc] init];
+        [pathRunnerRequest setSource:startRunnerItem];
+        [pathRunnerRequest setDestination:destinationItem];
+        [pathRunnerRequest setTransportType:MKDirectionsTransportTypeWalking];
+        [pathRunnerRequest setRequestsAlternateRoutes:NO];
+        
+        MKDirections *pathInterceptor = [[MKDirections alloc] initWithRequest:pathInterceptorRequest];
+        MKDirections *pathRunner = [[MKDirections alloc] initWithRequest:pathRunnerRequest];
+        [pathInterceptor calculateETAWithCompletionHandler:^(MKETAResponse * _Nullable interceptorResponse, NSError * _Nullable error1) {
+            if (interceptorResponse) {
```

Comment:
you should `dispatch_async(serialQueue` all of the logic in this nested completion block, so that your operations against `etasDifferences` and `etaPointPairs` are thread safe